payload/reggae/dub/info.d

```diff
@@ -196,17 +196,22 @@ struct DubInfo {
 
         auto packageTargets = targetsFunc()(files, flags, importPaths, stringImportPaths, [], projDir);
 
+        const dubPkgRoot = buildPath(dubPackage.path).deabsolutePath.stripRight(dirSeparator);
+
         // adjust object file output paths for all dub projects
         // optionally put the object files in dubObjsDir
         if(dubObjsDir.globalDir != "") {
+            const shortenedRoot = buildPath(dubObjsDir.globalDir, baseName(dubPackage.path));
             foreach(ref target; packageTargets) {
-                target.rawOutputs[0] = buildPath(dubObjsDir.globalDir,
-                                                options.projectPath.deabsolutePath,
-                                                dubObjsDir.targetDir,
-                                                target.rawOutputs[0]);
+                import std.digest.sha : sha256Of, toHexString;
+
+                const cmd = target.shellCommand(options);
+                //import std.stdio; writeln(".: ", cmd);
+                const hashstr = sha256Of(cmd).toHexString();
```

Comment:
The hash needs work - simply hashing the original cmdline includes the original output path etc., so that the hash diverges for each reggae build directory etc. It should also contain the used D compiler and its version.

Reply:
Hmm nope the vanilla output paths should be fine - they are independent from the reggae build dir and look like `.reggae/dcompile --objFile=home/mkinkelin/.dub/packages/arsd-official-10.9.8/arsd-official/minigui_addons_datetime_picker.a …` (so absolute without leading `/`). The source file paths seem to be absolute too.

The problem (different hashes for equivalent compilations) seems to mainly stem from the `-I` import paths - reggae apparently uses *all* dirs, for all compilations, not just the ones from dependencies of the dub package being compiled. So e.g. for an arsd subpackage, there's not a single `-I` as with dub itself, but 129 when compiling Symmetry's REPL, and 173 when compiling the full REPL.

Reply:
Do you mean hashing the command line as is as a string? That's not ideal as the hashing should be independent on spacing and ordering of flags etc. It's much better to hash the structured representation as a result of the decoding of the command line. This is what dub does. FYI, @atilaneves.

Reply:
Flags cannot be ordered in general. Try ordering `-flto=thin -flto=full -link-defaultlib-shared=false -link-defaultlib-shared` without changing the semantics...